core/arch/arm/plat-sunxi/main.c

```diff
@@ -171,3 +195,42 @@ void plat_cpu_reset_late(void)
 
 	tzpc_init();
 }
+#endif
+
+/*
+ * Allwinner's A64 has TZC380 like controller called SMC that can
+ * be programmed to protect parts of DRAM from non-secure world.
+ */
+#ifdef CFG_TZC380
+vaddr_t smc_base(void)
+{
+	return (vaddr_t)phys_to_virt(SUNXI_SMC_BASE, MEM_AREA_IO_SEC);
+}
+
+static TEE_Result smc_init(void)
+{
+	uint32_t val = 0;
+	vaddr_t base = smc_base();
+
+	if (!base) {
+		EMSG("smc not mapped");
+		panic();
+	}
+
+	tzc_init(base);
+	tzc_configure_region(0, 0x0, TZC_ATTR_REGION_SIZE(TZC_REGION_SIZE_1G) |
+			     TZC_ATTR_REGION_EN_MASK | TZC_ATTR_SP_ALL);
+	tzc_configure_region(1, 0x0, TZC_ATTR_REGION_SIZE(TZC_REGION_SIZE_32M) |
+			     TZC_ATTR_REGION_EN_MASK | TZC_ATTR_SP_S_RW);
+
+
+	/* SoC specific bits */
+	read32(base + SMC_MASTER_BYPASS);
+	val = val & ~(SMC_MASTER_BYPASS_EN_MASK);
+	write32(val, (base + SMC_MASTER_BYPASS));
+
```

Comment:
Sure we don't need a `dsb()` or such here?

Reply:
But it worked without need of dsb() or any delay.